### PR TITLE
make in-viewport compatible to ember-modifier v4

### PR DESCRIPTION
**Pull request by And012**
Closes # .

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,9 +24,12 @@
   "author": "Scott Newcomer",
   "license": "MIT",
   "dependencies": {
+    "@embroider/macros": "^1.8.3",
     "ember-auto-import": "^2.2.3",
     "ember-cli-babel": "^7.26.6",
-    "ember-modifier": "^2.1.2 || ^3.0.0",
+    "ember-destroyable-polyfill": "^2.0.3",
+    "ember-modifier": "^3.2.7",
```
Comment (SergeAstapov) [inline package.json:31]: @And012 we should change this to

```suggestion
    "ember-modifier": "^2.1.2 || ^3.0.0 || ^4.0.0",
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -61,6 +61,8 @@ jobs:
           - ember-default-with-jquery
           - embroider-safe
           - embroider-optimized
+          - ember-modifier-2.x
+          - ember-modifier-3.1
```
Comment (SergeAstapov) [inline .github/workflows/ci.yml:65]: @And012 this should include also scenarios `ember-modifier@3.2` and `ember-modifier@^4.0.0-beta.0` e.g. like its done in https://github.com/adopted-ember-addons/ember-keyboard/blob/master/.github/workflows/ci.yml#L133-L137

that way, we test compatibility with 2.x version, with 3.1 version and with 3.2 (where new API introduced and old API deprecated) as well as compatibility with 4.x where old API is removed

File: addon/modifiers/in-viewport.js [hunk 3]
```diff
@@ -4,92 +4,97 @@ import { inject as service } from '@ember/service';
...
-    );
-  }
-
-  @action
-  onEnter(...args) {
-    if (this.args.named.onEnter) {
-      this.args.named.onEnter.call(null, this.element, ...args);
+let modifier;
+
+if (macroCondition(dependencySatisfies('ember-modifier', '>=3.2.0 || 4.x'))) {
```
Comment (SergeAstapov) [inline addon/modifiers/in-viewport.js:14]: @And012 now it's missing `else` branch for compatibility with versions that provide old API

File: package.json [hunk 4]
```diff
@@ -24,9 +24,12 @@
...
+    "ember-modifier-manager-polyfill": "1.2.0",
```
Comment (SergeAstapov) [inline package.json:32]: @And012 we should not introduce `ember-modifier-manager-polyfill` as it's needed for Ember versions 3.8 and below and this addon is claim to be compatible with 3.16 and up

**Review by SergeAstapov — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 2 above)
Comment (sapryniukt) [reply .github/workflows/ci.yml:65]: @And012 also, pay attention to https://github.com/DockYard/ember-in-viewport/blob/master/config/ember-try.js
Here we need to declare each scenario that is added to `ci.yml`.

as a reference, the following scenarios for `ember-try.js` could look like
```javascript
  {
    name: 'ember-modifier-2.1',
    npm: {
      dependencies: {
        'ember-modifier': '^2.1.2',
      },
    },
  },
  {
    name: 'ember-modifier-3.1',
    npm: {
      dependencies: {
        'ember-modifier': '^3.1.0',
      },
    },
  },
  {
    name: 'ember-modifier-3.2',
    npm: {
      dependencies: {
        'ember-modifier': '^3.2.7',
      },
    },
  },
  {
    name: 'ember-modifier-4.0.0-beta.1',
    npm: {
      dependencies: {
        'ember-modifier': '^4.0.0-beta.1',
      },
    },
  }
```

**Comment by And012** [comment]
Have removed `ember-default-with-jquery` snenario.
`jquery-integration` optional feature flag to `true` was deprecated in Ember 3.x and removed in Ember 4.0.0.

File: package.json [hunk 5]
```diff
@@ -55,7 +57,7 @@
     "ember-qunit": "^5.1.5",
     "ember-resolver": "^8.0.3",
     "ember-responsive": "^4.0.2",
-    "ember-source": "~3.28.1",
+    "ember-source": "https://s3.amazonaws.com/builds.emberjs.com/release/shas/464e694afd611e2203759e5f76a14c7bfb023006.tgz",
```
Comment (SergeAstapov) [inline package.json:60]: @And012 @sapryniukt we can revert this, right?

File: package.json [hunk 6]
```diff
@@ -76,6 +78,9 @@
     "release-it-lerna-changelog": "^2.3.0",
     "webpack": "^5.58.2"
   },
+  "peerDependencies": {
+    "ember-source": "^3.8 || 4.*"
```
Comment (SergeAstapov) [inline package.json:82]: @And012 @sapryniukt why this is needed?

File: config/ember-try.js [hunk 7]
```diff
@@ -63,15 +63,34 @@ module.exports = async function () {
...
+      {
+        name: 'ember-modifier@3.2',
+        npm: {
+          dependencies: {
+            'ember-modifier': '^3.2.7',
+          },
+        },
+      },
+      {
+        name: 'ember-modifier@^4.0.0-beta.0',
```
Comment (SergeAstapov) [inline config/ember-try.js:90]: @And012 let's rename this to `ember-modifier@^4.0.0-beta.1` as there was no beta0 published to npm (and don';t forget to update `.github/workflows/ci.yml`)

File: package.json (hunk 6 above)
Comment (And012) [reply package.json:82]: `node_modules/@ember/render-modifiers/package.json`
"ember-source": "^3.8 || 4" here is defined likes so.
Cause of this in local environment all scenarious that need ember-source above 4. 
CI was not happy and was collapsing.
Screen Shot 2022-09-09 at 18 41 18

Comment (And012) [reply package.json:82]: during ember try:each it is added but I think I could remove it also

**Comment by SergeAstapov** [comment]
@cibernox @snewcomer would you be able to review/land this?

**Comment by miguelcobain** [comment]
Would love to get this fixed to get rid of deprecations. 👀

**Review by snewcomer — APPROVED** [review]
Nice changes 👍

**Comment by miguelcobain** [comment]
When can we expect a release? 🙏

**Comment by miguelcobain** [comment]
@snewcomer can we release, please?

**Pull request merged**